src/server/pkg/storage/chunker/chunker.go

```diff
@@ -0,0 +1,69 @@
+package chunker
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/chmduquesne/rollinghash/buzhash64"
+)
+
+const (
+	// MB is Megabytes.
+	MB = 1024 * 1024
+	// MinSize is the minimum chunk size.
+	MinSize = 5 * MB
+	// AverageBits determines the average chunk size (2^AverageBits).
+	AverageBits = 23
+	// MaxSize is the maximum chunk size.
+	MaxSize = 15 * MB
```

Comment:
It occurred to me that having a max chunk size is going to make rechunking after an insertion a lot more difficult because boundaries can shift due to an insertion. In the most degenerate case, where all of the chunks are of length `MaxSize` you'll have to rechunk the entire object. I'm not sure but I think that might be more trouble, both in terms of the code you have to write and the performance penalty, than tolerating large chunks in the case where we don't hit a content defined boundary.

Reply:
We are on the same page here, I realized this yesterday lol. There is actually an even bigger problem with using MaxSize in that it basically removes the guarantee that our file hashes are correct. The current thinking was that we would not use a chunk in a file hash unless it was fully contained within the file. Well, you can have a situation where the first split point in a file was determined by the MaxSize rather than the bit pattern which means that the hash of a file is influenced by the content of a prior file. Also, the MaxSize makes it difficult to parallelize this process because stitching together the pieces could involve shifting a lot of the split points. If we throw away the Min/Max Size constraints, we can parallelize this much easier and be confident in the file hashes, but we open ourselves up to certain pathological workloads. There might be some ways to handle this (a more/less restrictive bit pattern), but I have not really thought them through yet.

Reply:
I'm tempted to say we just throw out Min/Max sizing, do the simple thing for now and kind of punt on the pathological case. We might not ever be able to solve that case, it seems like we either have content defined chunking, or we do not, and if we do then it means that there's going to be some content that chunks badly and has really big, or really small, chunks. But it should be rare (probably very rare). We should also think about how bad these cases are, a big chunk means we'll have to download a lot of data, which could be a problem in terms of how much data we store in memory (although I think that can be some what alleviated with a streaming model for larger chunks) it also can create extra network traffic when we need to insert into the middle of that chunk. I'm not quite sure how big of a deal that would be, but it doesn't seem like it'd be a disaster.

Also, it seems like having a min size is still ok, because we'd still have all our split points at content defined splits, it just means that when we insert into the middle of that chunk, we might create some new split points by getting them over the size threshold, but that doesn't seem like extra complexity because we also might create some new splits by adding content with a split point. So we basically need to rechunk that chunk anyways.

Reply:
I agree that we might need to get rid of Min/Max sizes. Min sizes are still a problem in terms of file hashes because split points at the beginning of files with the same content could be different (a prior file with a split point closer to the end could cause the next split point to be later, while a prior file with a split point closer to the beginning could cause the split point in the next file to be earlier).